### PR TITLE
Use tags for TS metadata

### LINKED ISSUE #665: Use tags for TS metadata
I'm not sure if this is not possible due to a limitation of TS3, haven't looked into that.

Right now you have to disable TFAR when using ACRE, and vice versa. TFAR is already using a tag system for the metadata. `<TFAR>datas</TFAR>`. Would it be possible for ACRE to use `<ACRE>datas</ACRE>`

This would allow both plugins to be active at the same time.

### PR DESCRIPTION
**Pull request by BrettMayson**
**When merged this pull request will:**

- Closes #665
- Use C++17

Tested in MP with both TFAR and ACRE plugins enabled at the same time.

Replacement for PR #686

### CODE REVIEW

**Comment by BrettMayson** [comment]
Requesting review from @dedmen on the use of `string_view`

File: extensions/src/ACRE2TS/TsCallbacks_static.cpp [hunk 1]
```diff
@@ -34,6 +34,8 @@ void ts3plugin_infoData(uint64 serverConnectionHandlerID, uint64 id, enum Plugin
 
     BOOL noAcre = FALSE;
     char *metaData;
+    std::string result;
```
Comment (TheMagnetar) [inline extensions/src/ACRE2TS/TsCallbacks_static.cpp]: These should be  declared in the lowerest possible scope. In this case it is insinde the `if (ts3Functions.getClientVariableAsString(serverConnectionHandlerID, (anyID)id, CLIENT_META_DATA, &metaData) == ERROR_ok` statement

File: extensions/src/ACRE2TS/TsCallbacks_static.cpp [hunk 2]
```diff
@@ -34,6 +34,8 @@ void ts3plugin_infoData(uint64 serverConnectionHandlerID, uint64 id, enum Plugin
...
+    std::string sharedMsg;
```
Comment (TheMagnetar) [inline extensions/src/ACRE2TS/TsCallbacks_static.cpp]: same as above

File: extensions/src/ACRE2TS/TsCallbacks_static.cpp [hunk 3]
```diff
@@ -42,19 +44,24 @@ void ts3plugin_infoData(uint64 serverConnectionHandlerID, uint64 id, enum Plugin
...
-                int metaDataLength = strlen(metaData);
-                if (metaDataLength < 3 || metaDataLength > (INFODATA_BUFSIZE - 2)) {
+                if (!metaData) {
                     noAcre = TRUE;
+                } else {
+                    sharedMsg = metaData;
+                    size_t start_pos = sharedMsg.find(START_DATA);
+                    size_t end_pos = sharedMsg.find(END_DATA);
+                    if (start_pos == std::string::npos || end_pos == std::string::npos) {
+                        noAcre = TRUE;
```
Comment (TheMagnetar) [inline extensions/src/ACRE2TS/TsCallbacks_static.cpp]: Do not use upper case in here. Use `true` or `false`. This is an ongoing cleanup process.

File: extensions/src/ACRE2TS/TsCallbacks_static.cpp [hunk 4]
```diff
@@ -42,19 +44,24 @@ void ts3plugin_infoData(uint64 serverConnectionHandlerID, uint64 id, enum Plugin
...
-                if (!metaData) { 
-                    noAcre = TRUE;
-                }
-                int metaDataLength = strlen(metaData);
-                if (metaDataLength < 3 || metaDataLength > (INFODATA_BUFSIZE - 2)) {
+                if (!metaData) {
                     noAcre = TRUE;
+                } else {
+                    sharedMsg = metaData;
+                    size_t start_pos = sharedMsg.find(START_DATA);
```
Comment (TheMagnetar) [inline extensions/src/ACRE2TS/TsCallbacks_static.cpp]:
```suggestion
                    const size_t start_pos = sharedMsg.find(START_DATA);
```

File: extensions/src/ACRE2TS/TsCallbacks_static.cpp [hunk 5]
```diff
@@ -42,19 +44,24 @@ void ts3plugin_infoData(uint64 serverConnectionHandlerID, uint64 id, enum Plugin
...
+                    size_t end_pos = sharedMsg.find(END_DATA);
```
Comment (TheMagnetar) [inline extensions/src/ACRE2TS/TsCallbacks_static.cpp]:
```suggestion
                    const size_t end_pos = sharedMsg.find(END_DATA);
```

File: extensions/src/ACRE2TS/TsCallbacks_static.cpp [hunk 6]
```diff
@@ -42,19 +44,24 @@ void ts3plugin_infoData(uint64 serverConnectionHandlerID, uint64 id, enum Plugin
...
+                    if (start_pos == std::string::npos || end_pos == std::string::npos) {
+                        noAcre = TRUE;
+                    } else {
+                        result = sharedMsg.substr(start_pos + strlen(START_DATA), end_pos - start_pos - strlen(START_DATA));
+                    }
                 }
-                *data = (char*)malloc(INFODATA_BUFSIZE * sizeof(char)); 
+                *data = (char*)malloc(INFODATA_BUFSIZE * sizeof(char));
```
Comment (TheMagnetar) [inline extensions/src/ACRE2TS/TsCallbacks_static.cpp:59]: what happens if malloc fails to allocate the data? `data` should be checked against nullptr before proceeding with the rest of the function

File: extensions/src/ACRE2TS/TsCallbacks_static.cpp [hunk 7]
```diff
@@ -42,19 +44,24 @@ void ts3plugin_infoData(uint64 serverConnectionHandlerID, uint64 id, enum Plugin
...
-                }
-                int metaDataLength = strlen(metaData);
-                if (metaDataLength < 3 || metaDataLength > (INFODATA_BUFSIZE - 2)) {
+                if (!metaData) {
                     noAcre = TRUE;
+                } else {
+                    sharedMsg = metaData;
+                    size_t start_pos = sharedMsg.find(START_DATA);
+                    size_t end_pos = sharedMsg.find(END_DATA);
+                    if (start_pos == std::string::npos || end_pos == std::string::npos) {
```
Comment (TheMagnetar) [inline extensions/src/ACRE2TS/TsCallbacks_static.cpp]:
```suggestion
                    if ((start_pos == std::string::npos) || (end_pos == std::string::npos)) {
```

File: extensions/src/ACRE2TS/TS3Client.cpp [hunk 8]
```diff
@@ -117,6 +116,28 @@ acre::Result CTS3Client::exPersistVersion( void ) {
...
+    ts3Functions.getClientID(ts3Functions.getCurrentServerConnectionHandlerID(), &myID);
+    ts3Functions.getClientVariableAsString(ts3Functions.getCurrentServerConnectionHandlerID(), myID, CLIENT_META_DATA, &clientInfo);
+	std::string to_set;
+	std::string_view sharedMsg = clientInfo;
+    size_t start_pos = sharedMsg.find(START_DATA);
+    size_t end_pos = sharedMsg.find(END_DATA);
+	if (start_pos == std::string::npos || end_pos == std::string::npos) {
+		to_set = to_set + START_DATA + data + END_DATA;
+	} else {
+		std::string before = (std::string)sharedMsg.substr(0, start_pos);
```
Comment (TheMagnetar) [inline extensions/src/ACRE2TS/TS3Client.cpp]:
```suggestion
		const std::string before = (std::string)sharedMsg.substr(0, start_pos);
```

File: extensions/src/ACRE2TS/TS3Client.cpp [hunk 9]
```diff
@@ -117,6 +116,28 @@ acre::Result CTS3Client::exPersistVersion( void ) {
...
 }
 
+acre::Result CTS3Client::setSelfVariable(char* data) {
+    char* clientInfo;
+    anyID myID;
+    ts3Functions.getClientID(ts3Functions.getCurrentServerConnectionHandlerID(), &myID);
+    ts3Functions.getClientVariableAsString(ts3Functions.getCurrentServerConnectionHandlerID(), myID, CLIENT_META_DATA, &clientInfo);
+	std::string to_set;
+	std::string_view sharedMsg = clientInfo;
+    size_t start_pos = sharedMsg.find(START_DATA);
```
Comment (TheMagnetar) [inline extensions/src/ACRE2TS/TS3Client.cpp]:
```suggestion
    const size_t start_pos = sharedMsg.find(START_DATA);
```

File: extensions/src/ACRE2TS/TS3Client.cpp [hunk 10]
```diff
@@ -117,6 +116,28 @@ acre::Result CTS3Client::exPersistVersion( void ) {
...
+    size_t end_pos = sharedMsg.find(END_DATA);
```
Comment (TheMagnetar) [inline extensions/src/ACRE2TS/TS3Client.cpp]:
```suggestion
    const size_t end_pos = sharedMsg.find(END_DATA);
```

File: extensions/src/ACRE2TS/TS3Client.cpp [hunk 11]
```diff
@@ -117,6 +116,28 @@ acre::Result CTS3Client::exPersistVersion( void ) {
...
+	if (start_pos == std::string::npos || end_pos == std::string::npos) {
```
Comment (TheMagnetar) [inline extensions/src/ACRE2TS/TS3Client.cpp]:
```suggestion
	if ((start_pos == std::string::npos) || (end_pos == std::string::npos)) {
```

File: extensions/src/ACRE2TS/TS3Client.cpp [hunk 12]
```diff
@@ -117,6 +116,28 @@ acre::Result CTS3Client::exPersistVersion( void ) {
...
+		to_set = to_set + START_DATA + data + END_DATA;
+	} else {
+		std::string before = (std::string)sharedMsg.substr(0, start_pos);
+		std::string after = (std::string)sharedMsg.substr(end_pos + strlen(END_DATA), std::string::npos);
```
Comment (TheMagnetar) [inline extensions/src/ACRE2TS/TS3Client.cpp]:
```suggestion
		const std::string after = (std::string)sharedMsg.substr(end_pos + strlen(END_DATA), std::string::npos);
```

File: extensions/src/ACRE2TS/TS3Client.cpp [hunk 13]
```diff
@@ -117,6 +116,28 @@ acre::Result CTS3Client::exPersistVersion( void ) {
     return acre::Result::error;
 }
 
+acre::Result CTS3Client::setSelfVariable(char* data) {
```
Comment (TheMagnetar) [inline extensions/src/ACRE2TS/TS3Client.cpp]: I believe this is constant
```suggestion
acre::Result CTS3Client::setSelfVariable(const char *const data) {
```

File: extensions/src/ACRE2TS/TS3Client.cpp [hunk 14]
```diff
@@ -117,6 +116,28 @@ acre::Result CTS3Client::exPersistVersion( void ) {
...
+		to_set = to_set + START_DATA + data + END_DATA;
+	} else {
+		std::string before = (std::string)sharedMsg.substr(0, start_pos);
+		std::string after = (std::string)sharedMsg.substr(end_pos + strlen(END_DATA), std::string::npos);
+		to_set = before + START_DATA + data + END_DATA + after;
+	}
+    ts3Functions.setClientSelfVariableAsString(ts3Functions.getCurrentServerConnectionHandlerID(), CLIENT_META_DATA, to_set.c_str());
+	ts3Functions.freeMemory(clientInfo);
+	ts3Functions.flushClientSelfUpdates(ts3Functions.getCurrentServerConnectionHandlerID(), NULL);
+    return acre::Result::ok;
```
Comment (TheMagnetar) [inline extensions/src/ACRE2TS/TS3Client.cpp:137]: There seems to be some indentation discrepancies.

**Review by TheMagnetar — CHANGES_REQUESTED** [review]

File: extensions/src/ACRE2TS/TsCallbacks_static.cpp [hunk 15]
```diff
@@ -42,20 +42,30 @@ void ts3plugin_infoData(uint64 serverConnectionHandlerID, uint64 id, enum Plugin
...
+                    } else {
+                        result = sharedMsg.substr(start_pos + strlen(START_DATA), end_pos - start_pos - strlen(START_DATA));
+                    }
+                }
+                *data = (char*)malloc(INFODATA_BUFSIZE * sizeof(char));
+                if (data != nullptr) {
+                    if (!noAcre) {
+                        _snprintf_s(*data, INFODATA_BUFSIZE, INFODATA_BUFSIZE, "%s\n", result.c_str());
+                        ts3Functions.freeMemory(metaData);
+                    }
```
Comment (TheMagnetar) [inline extensions/src/ACRE2TS/TsCallbacks_static.cpp]: } else {

**Review by TheMagnetar — CHANGES_REQUESTED** [review]

File: extensions/src/ACRE2Arma/common/easylogging++.h [hunk 16]
```diff
@@ -15,7 +15,7 @@
 //
 #ifndef EASYLOGGINGPP_H
 #define EASYLOGGINGPP_H
-// Compilers and C++0x/C++11 Evaluation
+// Compilers and C++0x/c++17 Evaluation
```
Comment (dedmen) [inline extensions/src/ACRE2Arma/common/easylogging++.h]: did you actually update easylogging??? Doesn't look like it, there are no changes besides string replacements in comments?

File: extensions/src/ACRE2Arma/common/glm/detail/setup.hpp [hunk 17]
```diff
@@ -376,9 +376,9 @@
 #            define GLM_LANG GLM_LANG_CXX14
 #        elif __has_feature(cxx_decltype_auto) && __has_feature(cxx_aggregate_nsdmi) // GLM_COMPILER_LLVM33 + -std=c++1y
 #            define GLM_LANG GLM_LANG_CXX1Y
-#        elif __cplusplus >= 201103L // GLM_COMPILER_LLVM33 + -std=c++11
+#        elif __cplusplus >= 201103L // GLM_COMPILER_LLVM33 + -std=c++17
```
Comment (dedmen) [inline extensions/src/ACRE2Arma/common/glm/detail/setup.hpp]: No that's not what this code does. This code checks for c++11 not 17.

File: extensions/src/ACRE2Arma/common/glm/detail/setup.hpp [hunk 18]
```diff
@@ -653,7 +653,7 @@
 #endif
 
 // OpenMP
-#ifdef _OPENMP 
+#ifdef _OPENMP
```
Comment (dedmen) [inline extensions/src/ACRE2Arma/common/glm/detail/setup.hpp:656]: Why are all these space at end of line changes part of this PR, they have nothing to do with this

File: extensions/src/ACRE2Arma/common/glm/gtx/range.hpp [hunk 19]
```diff
@@ -44,7 +44,7 @@
 #include "../detail/setup.hpp"
 
 #if !GLM_HAS_RANGE_FOR
-#    error "GLM_GTX_range requires C++11 suppport or 'range for'"
+#    error "GLM_GTX_range requires c++17 suppport or 'range for'"
```
Comment (dedmen) [inline extensions/src/ACRE2Arma/common/glm/gtx/range.hpp]: What? range for is c++11 not 17.

File: extensions/src/ACRE2Arma/common/glm/gtx/scalar_multiplication.hpp [hunk 20]
```diff
@@ -44,7 +44,7 @@
 #include "../detail/setup.hpp"
 
 #if !GLM_HAS_TEMPLATE_ALIASES && !(GLM_COMPILER & GLM_COMPILER_GCC)
-#    error "GLM_GTX_scalar_multiplication requires C++11 suppport or alias templates and if not support for GCC"
+#    error "GLM_GTX_scalar_multiplication requires c++17 suppport or alias templates and if not support for GCC"
```
Comment (dedmen) [inline extensions/src/ACRE2Arma/common/glm/gtx/scalar_multiplication.hpp]: No it doesn't.

File: extensions/src/ACRE2TS/TS3Client.cpp [hunk 21]
```diff
@@ -117,6 +115,28 @@ acre::Result CTS3Client::exPersistVersion( void ) {
     return acre::Result::error;
 }
 
+acre::Result CTS3Client::setSelfVariable(const char *const data) {
```
Comment (dedmen) [inline extensions/src/ACRE2TS/TS3Client.cpp]: "SelfVariable" == "ClientMetaData" ?
Doesn't seem that readable/conclusive to me.

File: extensions/src/ACRE2Arma/common/glm/detail/setup.hpp (hunk 18 above)
Comment (BrettMayson) [reply extensions/src/ACRE2Arma/common/glm/detail/setup.hpp:656]: VS Code is removing them when I save a file

File: extensions/src/ACRE2Arma/common/easylogging++.h [hunk 22]
```diff
@@ -320,7 +320,7 @@ ELPP_INTERNAL_DEBUGGING_OUT_INFO << ELPP_INTERNAL_DEBUGGING_MSG(internalInfoStre
 #  define ELPP_VERBOSE_LOG 0
 #endif  // (!defined(ELPP_DISABLE_VERBOSE_LOGS) && (ELPP_LOGGING_ENABLED))
 #if (!(ELPP_CXX0X || ELPP_CXX11))
-#   error "C++0x (or higher) support not detected! (Is `-std=c++11' missing?)"
+#   error "C++0x (or higher) support not detected! (Is `-std=C++11' missing?)"
```
Comment (jonpas) [inline extensions/src/ACRE2Arma/common/easylogging++.h]:
```suggestion
#   error "C++0x (or higher) support not detected! (Is `-std=c++11' missing?)"
```

File: extensions/src/ACRE2Arma/common/glm/detail/setup.hpp [hunk 23]
```diff
@@ -477,7 +477,7 @@
 #    elif GLM_LANG & GLM_LANG_CXX1Y_FLAG
 #        pragma message("GLM: C++1y")
 #    elif GLM_LANG & GLM_LANG_CXX11_FLAG
-#        pragma message("GLM: C++11")
+#        pragma message("GLM: c++17")
```
Comment (jonpas) [inline extensions/src/ACRE2Arma/common/glm/detail/setup.hpp]:
```suggestion
#        pragma message("GLM: C++11")
```

File: extensions/src/ACRE2Arma/common/glm/detail/setup.hpp (hunk 18 above)
Comment (jonpas) [reply extensions/src/ACRE2Arma/common/glm/detail/setup.hpp:656]: Spaces are fine, need to clean that up anyways.

EDIT: This was supposed to be a comment to above comment...

**Review by jonpas — CHANGES_REQUESTED** [review]

**Review by TheMagnetar — APPROVED** [review]

File: extensions/src/ACRE2TS/TS3Client.h [hunk 24]
```diff
@@ -62,8 +62,8 @@ class CTS3Client: public IClient {
 
     std::string getTempFilePath( void );
     std::string getConfigFilePath(void);
-
-    acre::Result playSound(std::string path_, acre::Vector3_t position_, const float32_t volume_, const int32_t looping_);
+
```
Comment (jonpas) [inline extensions/src/ACRE2TS/TS3Client.h]: Unneeded indentation.

**Comment by jonpas** [comment]
Needs to set C++17 for MSVC as well.

**Comment by jonpas** [comment]
Added C++17 for MSVC in CMake. Only for ACRE2TS project as some things are not fully compatible or CMake setup is pretty old and would require further changes.

This does bring a deprecation warning of `concurrent_unordered_set`'s `allocator` members which are deprecated. Should use `allocator_traits` with C++17, but it works for now.

**Review by jonpas — APPROVED** [review]

**Pull request merged**